src/notifications/signal.h

```diff
@@ -0,0 +1,111 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ */
+
+#ifndef WL_NOTIFICATIONS_SIGNAL_H
+#define WL_NOTIFICATIONS_SIGNAL_H
+
+#include <cassert>
+#include <functional>
+#include <list>
+#include <memory>
+#include <set>
+
+#include "base/macros.h"
+#include "base/wexception.h"
+
+namespace Notifications {
+
+/** Class that allows attaching callback functions to an object. */
+template <typename... Args> class Signal {
+public:
+	/** Wrapper around a callback function. */
+	struct SignalSubscriber {
+		explicit SignalSubscriber(const Signal& p, const std::function<void(Args...)>& c)
+		   : parent_(p), callback_(c) {
+		}
+		~SignalSubscriber() {
+			parent_.unsubscribe(this);
+		}
+
+		const Signal& parent_;
+		const std::function<void(Args...)> callback_;
+
+		DISALLOW_COPY_AND_ASSIGN(SignalSubscriber);
+	};
+
+	/** Invoke all the signal's subscribers' callback functions. */
+	void operator()(Args... args) const {
+		for (const auto& s : all_subscribers_) {
+			s->callback_(args...);
+		}
+	}
+
+	/** Create a subscriber with a user-defined lifetime. */
+	std::unique_ptr<SignalSubscriber> subscribe(const std::function<void(Args...)>& callback,
+	                                            bool at_front = false) const {
+		SignalSubscriber* s = new SignalSubscriber(*this, callback);
+		if (at_front) {
+			all_subscribers_.push_front(s);
+		} else {
+			all_subscribers_.push_back(s);
+		}
+		return std::unique_ptr<SignalSubscriber>(s);
+	}
+
+	/** Create a subscriber with the same lifetime as the signal. */
+	inline void connect(const std::function<void(Args...)>& callback, bool at_front = false) const {
+		owned_subscribers_.insert(subscribe(callback, at_front));
+	}
+
+	/** Create a subscriber that echoes the signal's invokations to another signal. */
+	inline std::unique_ptr<SignalSubscriber> subscribe(const Signal& s,
+	                                                   bool at_front = false) const {
+		return subscribe([&s](Args... args) { s(args...); }, at_front);
+	}
+	inline void connect(const Signal& s, bool at_front = false) const {
+		connect([&s](Args... args) { s(args...); }, at_front);
+	}
+
+	/** Called by a subscriber at the end of its lifetime. */
+	void unsubscribe(const SignalSubscriber* s) const {
+		for (auto it = all_subscribers_.begin(); it != all_subscribers_.end(); ++it) {
+			if (*it == s) {
+				all_subscribers_.erase(it);
+				return;
+			}
+		}
+		NEVER_HERE();
+	}
+
+	Signal() = default;
+	~Signal() {
+		owned_subscribers_.clear();
+		// Any subscribers not owned by us should have been destroyed by their owner by now
+		assert(all_subscribers_.empty());
+	}
+
+private:
+	mutable std::set<std::unique_ptr<SignalSubscriber>> owned_subscribers_;
```

Comment:
why is this mutable needed?

Reply:
A `Signal` instance can sometimes be a member of a `const` object, but we still want to allow new subscriptions to the object, so all member functions of `Signal` need to be `const`-qualified while still being able to add/remove subscribers to/from these lists.

Reply:
This is an age old debate - I remember having it at Google a few times. Should "const" mean - conceptually this is constant? or should it mean the bytes of this data structure do not change. We felt that using mutable let subtle bugs creep in, i.e. in our projects we opted for making classes non-const instead of using mutable. I do understand that this would introduce a much large change in the code base. I suggest to add a comment here that basically said what you explained me in writing.

Reply:
I usually try to avoid `mutable` and make as many variables and functions `const` as possible, but to me `const` means that "real" properties of an object remain the same whereas additional, let's call it "metadata", may still change, and this is such a case. For those cases `mutable` is acceptable IMHO (better than a `const_cast` anyway).

That said, a `static std::list<std::pair<std::set<std::unique_ptr<SignalSubscriber>>, std::list<SignalSubscriber*>>>` (better with some named `struct`s and aliases) plus a pointer to an entry in the `set` as a member of the `Signal` class would achieve the same thing without using `mutable`, though might be slightly less readable. Would this be preferable?

Reply:
no, I think mutable is better than a static here. In general I also agree with your thoughts about mutable, especially here not allowing it and avoiding const casts would have a large ripple effect through the code base. I just ask for a comment to explain.